lib/ZumoOta/src/BootloaderCom.cpp

```diff
@@ -95,28 +103,246 @@ class IntProgCmdProvider: public CmdProvider {
+     * memory address for sequential data writes.
+     */
+    uint16_t  m_currWriteMemAddr; 
+
+   /**
+    * @brief Updates the command.
+    * This method updates the provided command based on specific requirements
+    * before it is sent for execution. It modifies the command buffer and size
+    * to reflect the changes required by the bootloader communication protocol.
+    * @param fwProgCmd Pointer to the CommandInfo structure representing the command to be updated.
+    * The method modifies the command buffer and size to reflect the changes.
+    * @remarks This method is called before sending each command to the bootloader for execution.
+    * The update process includes modifying command identifiers and parameters to adhere to the
+    * bootloader communication protocol requirements.
+    */
+    bool updateFwProgCmd(CommandInfo& fwProgCmd)
+    {
+        if (m_nextCmd == FLSPRG_SET_ADDR)
+        {
+            /*Perform the required updates.*/
+            m_updatedCmdBuffer[0] = 0x41;  /**< The first value is always 0x41.*/
+            m_updatedCmdBuffer[1] = (m_currWriteMemAddr >> 8) & 0xFF; /**<Higher 8 bits of m_currWriteMemAddr.*/
+            m_updatedCmdBuffer[2] = m_currWriteMemAddr & 0xFF; //**<Lower 8 bits of m_currWriteMemAddr.*/
+            /*Increase m_currWriteMemAddr by 64 for the next command.*/
+            m_currWriteMemAddr += 64U;
+            /**<Set the Cmd instance to the buffer.*/
+            fwProgCmd.command = m_updatedCmdBuffer;
+            fwProgCmd.commandsize = 3;
+            m_nextCmd = FLSPRG_WRITE_BLOCK;
+            return true;
+        }
+        else if (m_nextCmd == FLSPRG_WRITE_BLOCK)
+        {
+            /*Update the command identifier bytes for write block*/
+            m_updatedCmdBuffer[0] = 0x42;
+            m_updatedCmdBuffer[1] = 0x00;
+            m_updatedCmdBuffer[2] = 0x80;
+            m_updatedCmdBuffer[3] = 0x46;
+            /*Read the file contents into m_firmwareBuffer for storage.*/
+            m_firmwareBytesRead = m_firmwareFile.read(
+                m_firmwareBuffer,
+                FLASH_BLOCK_LENGTH);
+            /*Check if there was an error reading the file.*/
+            if (m_firmwareBytesRead < 0)
+            {
+                /*Error reading the file.*/
+                LOG_ERROR("No Bytes available!");
+                return false;
+            } 
+            else if (0 == m_firmwareBytesRead)
+            {
+                m_nextCmd = FLSPRG_COMPLETE;
+                return false;
+            }
+            /*If the number of bytes read is less than MAX_BYTES_TO_READ,fill the rest of the buffer with a pattern.*/
+            size_t gapFill = FLASH_BLOCK_LENGTH - m_firmwareBytesRead;
+            if (0U != gapFill)
+            {
+                m_nextCmd = FLSPRG_COMPLETE;
+                while (gapFill > 0U)
+                {
+                    m_firmwareBuffer[m_firmwareBytesRead] = 0xFF;

```

Comment:
again use m_updatedCmdBuffer[WRITE_BLOCK_CMD_LENGTH + m_firmwareBytesRead]

Reply:
We can't use WRITE_BLOCK_CMD_LENGTH + m_firmwareBytesRead. It would create a buffer with a length of 132, which is not what we want. Instead, I used m_updatedCmdBuffer[m_firmwareBytesRead].